flink-java/src/main/java/org/apache/flink/api/java/utils/ParameterTool.java

```diff
@@ -68,81 +68,46 @@
-					// we got into a new key, even though we were a value --> current key is one without value
-					if (value != null) {
-						throw new IllegalStateException("Unexpected state");
-					}
-					map.put(key, NO_VALUE_KEY);
-					// key will be overwritten in the next step
-				}
-				key = arg.substring(2);
-				expectValue = true;
-			} // check for - argument
-			else if (arg.startsWith("-")) {
-				// we are waiting for a value, so this is a - prefixed value (negative number)
-				if (expectValue) {
-
-					if (NumberUtils.isNumber(arg)) {
-						// negative number
-						value = arg;
-						expectValue = false;
-					} else {
-						if (value != null) {
-							throw new IllegalStateException("Unexpected state");
-						}
-						// We waited for a value but found a new key. So the previous key doesnt have a value.
-						map.put(key, NO_VALUE_KEY);
-						key = arg.substring(1);
-						expectValue = true;
-					}
-				} else {
-					// we are not waiting for a value, so its an argument
-					key = arg.substring(1);
-					expectValue = true;
-				}
+		final Map<String, String> map = new HashMap<>(args.length / 2);
+
+		final String errorMessage = "Error parsing arguments '%s' on '%s'.";
+
+		int i = 0;
+		while (i < args.length) {
+			final String key;
+
+			if (args[i].startsWith("--")) {
+				key = args[i].substring(2);
+			} else if (args[i].startsWith("-")) {
+				key = args[i].substring(1);
 			} else {
-				if (expectValue) {
-					value = arg;
-					expectValue = false;
-				} else {
-					throw new RuntimeException("Error parsing arguments '" + Arrays.toString(args) + "' on '" + arg + "'. Unexpected value. Please prefix values with -- or -.");
-				}
+				throw new IllegalArgumentException(
+					String.format(errorMessage + " Please prefix keys with -- or -.",
+						Arrays.toString(args), args[i]));
 			}
 
-			if (value == null && key == null) {
-				throw new IllegalStateException("Value and key can not be null at the same time");
-			}
-			if (key != null && value == null && !expectValue) {
-				throw new IllegalStateException("Value expected but flag not set");
-			}
-			if (key != null && value != null) {
-				map.put(key, value);
-				key = null;
-				value = null;
-				expectValue = false;
-			}
-			if (key != null && key.length() == 0) {
-				throw new IllegalArgumentException("The input " + Arrays.toString(args) + " contains an empty argument");
+			if (key.length() == 0) {
+				throw new IllegalArgumentException(
+					String.format(errorMessage + " Keys should never be empty",
+						Arrays.toString(args), args[i]));
 			}
 
-			if (key != null && !expectValue) {
+			i += 1; // try to find the value
+
+			if (i >= args.length) {
+				map.put(key, NO_VALUE_KEY);
+			} else if (NumberUtils.isNumber(args[i])) {
+				map.put(key, args[i]);
+				i += 1;
+			} else if (args[i].startsWith("--")) {
+				map.put(key, NO_VALUE_KEY);
+			} else if (args[i].startsWith("-")) { // negate number guarded by precondition
```

Comment:
It's not clear to me what the comment means, and you have two equal branches. Maybe rewrite to:

```
			else if (args[i].startsWith("--") || args[i].startsWith("-")) {
				// the argument cannot be a negative number because we checked earlier
				// -> the next argument is a parameter name
				map.put(key, NO_VALUE_KEY);
			}
```

Reply:
Thanks for the advice, address it.